tests/test_directives.py

```diff
@@ -0,0 +1,118 @@
+        directive_copy.parallel = lambda x, y: x*y/(x+y)
+        assert directive_copy.parallel(3, 2) == 1.2
+        with pytest.raises(TypeError):
+            directive_copy.parallel('4', '2')
+        with pytest.raises(ZeroDivisionError):
+            directive_copy.parallel(0, 0)
+
+    def test_validation(self):
+        for directive in available_directives_list:
+            directive.validation(directive.default)
+
+    def test_validation_manual(self):
+        NP_copy = deepcopy(NP)
+        NP_copy.default = 0
+        with pytest.raises(ValueError):
+            # As 0 is not a natural number this should raise an error
+            NP_copy.validation(NP_copy.default)
+
+        def perf_square(v):
+            if (v**0.5) % 1 == 0:
+                return True
+            return False
+
+        NP_copy.validation = perf_square
+        assert NP_copy.validation(4)
+        assert not NP_copy.validation(5)
+        with pytest.raises(TypeError):
+            NP_copy.validation('4')
+
+    def test_default(self):
+        assert NP.default == 1
+        assert EXECUTABLE.default == sys.executable
+
+    def test_finalize(self):
+        # Yet to find an appropriate way to test this
+        assert True
+
+
+class TestDirectives:
+    """
+    Tests for Directives Class
+    """
+    def test_get_directives(self, setUp):
+        directives = setUp(available_directives_list=available_directives_list)
+        for item in available_directives_list:
+            assert directives[item.name] == item.default
+
+    def test_add_directives(self, setUp):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        directives._add_directive(OMP_NUM_THREADS)
+        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+        with pytest.raises(TypeError):
+            directives._add_directive('Test')
+        with pytest.raises(ValueError):
+            directives._add_directive(NP)
+
+    def test_set_directives(self, setUp):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        directives[OMP_NUM_THREADS.name] = OMP_NUM_THREADS.default
+        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+
+    def test_define_directive(self, setUp):
+        directives = setUp(available_directives_list=available_directives_list[:-1])
+        # The values below corresponds to NP, NRANKS, NGPU, EXECUTABLE.
+        valid_values = [100, 100, 100, 'My String']
+        invalid_values = [0, -1, -1, [0]]
+        for idx, dir in enumerate(available_directives_list[:-1]):
+            directives._set_defined_directive(dir.name, valid_values[idx])
+            assert directives[dir.name] == valid_values[idx]
+        for idx, dir in enumerate(available_directives_list):
+            with pytest.raises(ValueError):
+                directives._set_defined_directive(dir.name, invalid_values[idx])
+
+    def test_del_directive(self, setUp):
```

Comment:
This is on the right track, but I would rather not use a `_DirectivesItem` for the purposes of a user-directive. A simple name: value mapping is fine `['test'] = True` is fine.

Reply:
Okay, noted.